feat(Android,**unstable**): integration with `react-native-edge-to-edge`

zoontek opened the pull request:
> [!note]
> note by @kkafar: 
> Please note that I merge this as unstable behaviour, since I might want to remove the dependency & code once we land https://github.com/facebook/react-native/pull/47554 in core or endup with only edge-to-edge supported by Android.


## Description

This follow a discussion started [on the react-navigation repository](https://github.com/react-navigation/react-navigation/pull/12217).

## Motivation

The future of Android is [edge-to-edge](https://developer.android.com/about/versions/15/behavior-changes-15#edge-to-edge), and to make the React Native developer experience seamless in this regard, the ecosystem needs to transition from “opaque system bars by default” to “edge-to-edge by default.”

To prevent library authors from implementing their own edge-to-edge solutions—which could interfere with other libraries—and because it’s not possible to reliably detect if edge-to-edge is already enabled on Android, we have collaborated with [Expo](https://expo.dev/) to create a library that handles this functionality and is detectable using a simple helper: [`react-native-is-edge-to-edge`](https://github.com/zoontek/react-native-edge-to-edge/tree/main/react-native-is-edge-to-edge).

This approach allows you to bypass certain options and props (in this case, `Screen` `statusBarTranslucent`, `navigationBarTranslucent`, `statusBarColor` and `navigationBarColor` - [setting background color is deprecated](https://developer.android.com/about/versions/15/behavior-changes-15#deprecated-apis)). All are Android only props and are now obsolete / deprecated (at least, when running in edge-to-edge mode)

## Changes

- Add `react-native-is-edge-to-edge` to bypass `statusBarTranslucent`, `navigationBarTranslucent`, `statusBarColor` and `navigationBarColor` values
- Add a warning about their usage when running in edge-to-edge mode.

## Test code and steps to reproduce

- Install [react-native-edge-to-edge](https://github.com/zoontek/react-native-edge-to-edge) in the Example app.
- Force `react-navigation` to use local version of `react-native-screens`.

> [!WARNING]  
> For the `StatusBar` example, `react-native` `headerTopInsetEnabled` might interfere when `statusBarTranslucent` is explicitely set to `false`, but the user will receive a warning inviting him to remove the option / prop, so that's acceptable.

## More

If you want to bypass some Kotlin code directly, consider this util:

```kotlin
object EdgeToEdgeUtil {
  val ENABLED: Boolean
    get() = try {
      // we cannot detect edge-to-edge, but we can detect react-native-edge-to-edge install
      Class.forName("com.zoontek.rnedgetoedge.EdgeToEdgePackage")
      true
    } catch (exception: ClassNotFoundException) {
      false
    }
}
```

## Checklist

- [ ] Included code example that can be used to test this change
- [ ] Updated TS types
- [ ] Updated documentation: <!-- For adding new props to native-stack -->
  - [ ] https://github.com/software-mansion/react-native-screens/blob/main/guides/GUIDE_FOR_LIBRARY_AUTHORS.md
  - [ ] https://github.com/software-mansion/react-native-screens/blob/main/native-stack/README.md
  - [ ] https://github.com/software-mansion/react-native-screens/blob/main/src/types.tsx
  - [ ] https://github.com/software-mansion/react-native-screens/blob/main/src/native-stack/types.tsx
- [ ] Ensured that CI passes

kkafar commented:
Hey @zoontek I took first glance I want to reiterate to make sure I get this right: 

the goal here is to extract status bar / navigation bar management in terms of its visibility from all packages beside `react-native-is-edge-to-edge`, so that there is one, centralised solution they (other packages) can use to detect, whether the app is in edge-to-edge. Moreover -  manipulation of visibility of these elements should be done only through API of this package.

Note that changes you suggest are **breaking**.

I like the direction in general, assuming that we get broader adoption this might be a good thing. Do yo have a list of packages that need to undergo these changes prepared? If so, please share it. I'm asking for it, because it is related to my another point below 👇🏻 

I'm reluctant to introduce any third party dependency, especially one with native code. `react-native-screens` is a upstream package and such dependencies become liabilities especially when it comes to adopting to breaking changes rolled out in react-native between versions. We have this situation already partially (because only in example-apps, not in library directly) with different package, which requires us often to provide patches and upstream them, adding maintenance burden. Having this is mind I wonder whether it would be maybe better direction to get rid of this set of functionalities all together from `react-native-screens`? I haven't made my mind yet. I'm willing to change it. I'll forward this PR internally in order to get some comments on whether we should do it or not.

satya164 commented:
> I'm reluctant to introduce any third party dependency, especially one with native code.

afaik it's a JS only and lightweight dep that checks if the `react-native-edge-to-edge` module is available.

kkafar commented:
> As Android 15+ (targetting SDK 15) doesn't allow opaque system bars and enforces apps to draw under, the library enforces edge-to-edge and actually don't allow switching back to opaque bars anymore (because it will be impossible in the future - same for setting bars background color) - it doesn't want to handle their visibility, it aligns with (future) Android enforced behaviour.

I see, makes even more sense now. 

> I understand, but there's currently no native code involved in the PR, it just detect if another package is installed.

Yeah, I've missed the difference between `react-native-is-edge-to-edge` and `react-native-edge-to-edge` and inspected only the code of the latter. Dependency on `react-native-is-edge-to-edge` is much more acceptable.

> [...] we had only two options: a common library + lib detection or the feature being included in RN core (we asked for it, it might arrive at some point, but not fot the moment).

It would be good to upstream it to core at some point, definitely. I see your point though.


Thank you for detailed explanation! I'm leaning towards accepting this change, but need some more time to think this through (maybe we want to go more radically and get rid of some future-scope from this library entirely).

zoontek commented:
@kkafar If it helps take a decision, a new library added support: [react-native-bottom-tabs](https://okwasniewski.github.io/react-native-bottom-tabs/docs/guides/edge-to-edge-support.html) (which plays nice with `react-navigation)`

zoontek commented:
@kkafar FYI, we continue pushing for it. The react native team just merged [edge-to-edge for Modal into core](https://github.com/facebook/react-native/pull/47254#issuecomment-2461335718).

satya164 commented:
@zoontek I'm curious whether these also apply to React Native's `StatusBar` component/module since it also changes status bar's color/translucency like the code here.

satya164 commented:
@zoontek the other thing you mentioned here is that user needs to manually add these options until this PR is merged https://github.com/zoontek/react-native-edge-to-edge/issues/45#issuecomment-2472863904

But by default these values are `undefined` iirc, and they'll also be `undefined` after this PR is merged. So unless user is explicitly overriding somewhere, then there shouldn't be any change in the default behavior, right? Or was this a misunderstanding?

zoontek commented:
@satya164 Yes, there's no need to manually set them indeed, only to avoid that those are not set to `undefined`. My first though here was that something set`statusBarTranslucent` / `navigationBarTranslucent` to `false` somewhere in the dependencies chain or other were set to `false` somewhere (in `react-navigation`? in `expo-router`? I didn't checked yet at the time I wrote the reply), but the issue was with builtin `StatusBar` indeed.

If those stay `undefined`, there's indeed no issue. But it's very easy to imagine that a dependency, somewhare use:

```tsx
type Props = {
  statusBarTranslucent?: boolean
}

// Here, defaulting to `false` breaks edge-to-edge, it's vicious
const MyComponent = ({ statusBarTranslucent = false }: Props) => {
  return <Screen statusBarTranslucent={statusBarTranslucent} />
}
```

This kind of issues are super hard to debug if you don't understand the topic perfectly, so a nightmare for users that only want to enable it and stop thinking about it.

> @zoontek I'm curious whether these also apply to React Native's StatusBar component/module since it also changes status bar's color/translucency like the code here. Does it make sense to prevent the user from changing statusbar color if they enable edge to edge? What if they want to do it on one screen?

`StatusBar` is the elephant in the room. They put a warning about `backgroundColor` in the documentation:

<img width="1118" alt="Screenshot 2024-11-18 at 14 04 23" src="https://github.com/user-attachments/assets/107a7300-e764-488c-90cb-f2f4260e654c">

But there's still in the module [some code that breaks edge-to-edge](https://github.com/facebook/react-native/blob/v0.76.2/packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/view/WindowUtil.kt#L59): we saw it with the `expo-router` issue, that just mounted a `StatusBar` in `ExpoRoot`. For the moment, `react-native-edge-to-edge` `SystemBars` component is the safe way to go.

To tackle the edge-to-edge issue, Meta team asked me for a proposal, it started [here](https://github.com/facebook/react-native/pull/47554) (very WIP). The idea is similar:

- Having a way to enable edge-to-edge easily (`enableEdgeToEdge=true`)
- Give to library authors an easy way to detect if it's enabled in native code (`BuildConfig.IS_EDGE_TO_EDGE_ENABLED`)
- Give to non-native libraries authors a way to detect if it's enabled in JS (`Appearance.isEdgeToEdge()`)

And of course, this PR takes care of [fixing the StatusBar module behavior](https://github.com/facebook/react-native/pull/47554/files#diff-b56ad144ce3d7787344bc9a51c568eafaa0e377f193390a02739723ea0dfebcd).

zoontek commented:
> Does it make sense to prevent the user from changing statusbar color if they enable edge to edge?

For me, yes. As it will not work in the future (when targetSDK = 35 and on Android 15+):

<img width="656" alt="Screenshot 2024-11-18 at 15 17 55" src="https://github.com/user-attachments/assets/40f188af-6419-42d0-9298-3e89ed16a21d">

Note that the users can create a `<View />` with `height: insets.top` and a background color. This still works on Android 15+ and it also works on iOS.

satya164 commented:
I think if edge-to-edge will be part of core, in addition to these, it'll be better to mark these props as deprecated as well so they can be removed entirely in the future.

kkafar commented:
*Note: List of resources is for future me* 😅 

I've familiarised myself with pretty much all the resources you guys linked:

1. [Android behaviour changes when targeting API Level 35+](https://developer.android.com/about/versions/15/behavior-changes-15#window-insets)
2. @zoontek's [Proposal to core](https://github.com/facebook/react-native/pull/47554)
3. [Last year's timeline of enforcing minimal `targetSdk` for PlayStore apps](https://support.google.com/googleplay/android-developer/answer/11926878?sjid=11853000253346477363-EU#zippy=%2Care-there-any-exceptions-for-existing-apps-targeting-api-or-below)
4. Once again with all the discussion above ☝🏻 
5. Source code of both [`react-native-edge-to-edge` & `react-native-is-edge-to-edge`](https://github.com/zoontek/react-native-edge-to-edge?tab=readme-ov-file)

Conclusions: 

the arguments made here are compelling. We'll proceed with merging this PR, after I'm done with testing (should be able to do this by the end of the week). **My only concerns right now are Fabric-related**, as there is no such thing as `undefined` value for props on new architecture - codegened component's code always passes some non-null (non-undefined) default value, therefore we need to ensure that these defaults are appropriate. 

Another thing I want to address is handling of `topInsetEnabled` prop on `HeaderConfig` [(link)](https://github.com/react-navigation/react-navigation/blob/113cea5770fbca7cc86594d7dac1f00c2c392eec/packages/native-stack/src/views/NativeStackView.native.tsx#L229-L238). It would be nice if we had knowledge of whether the app is in edge-to-edge in react-navigation code, to be able to set this prop and prevent content jumping appropriately - I'm thinking whether we should not consider adding `react-native-is-edge-to-edge` as dependency downstream as well @satya164. Alternative approach would be to remove this `topInsetEnabled` property entirely - but this is not feasible before the changes land in core, I believe. 

WIP WIP (I ned to write few more paragraphs)

zoontek commented:
> My only concerns right now are Fabric-related, as there is no such thing as undefined value for props on new architecture - codegened component's code always passes some non-null (non-undefined) default value, therefore we need to ensure that these defaults are appropriate.

@kkafar  We could also update the code to be something like this:

```tsx
const edgeToEdgeProps = EDGE_TO_EDGE
  ? {
      navigationBarTranslucent: true,
      statusBarTranslucent: true,
    }
  : {
      ...(navigationBarColor != null && { navigationBarColor }),
      ...(navigationBarTranslucent != null && { navigationBarTranslucent }),
      ...(statusBarColor != null && { statusBarColor }),
      ...(statusBarTranslucent != null && { statusBarTranslucent }),
    };
    
// …

return <AnimatedScreen
  {...props}
  {...edgeToEdgeProps}
```

zoontek commented:
@kkafar @satya164 I rebased against main and updated the PR to avoid passing `statusBarTranslucent` / `navigationBarTranslucent` / `statusBarColor` / `navigationBarColor` completly (not even when `undefined`)

![Screenshot 2024-11-26 at 11 09 32](https://github.com/user-attachments/assets/283c8cd6-61ce-474a-a1e9-37e12c6ebb18)

zoontek commented:
Hi @kkafar, I hate to do this kind of comments, but do you have any news? 😅

kkafar commented:
I'll be rolling out these changes in either 4.7.0 beta or 4.8.0 beta, depending on few other PRs I want to land & having this change released in separation.

kkafar commented:
Also note for the future: I've decided on [deprecating all edge-to-edge related props](https://github.com/software-mansion/react-native-screens/pull/2638). They will be removed in future versions of screens.

zoontek commented:
@kkafar Perfect! I will rebase to resolve the conflict.

zoontek commented:
@kkafar Rebased! As there were some linting issues, I extracted the logic in a helper function. This way, it will have no extra cost for non-users of edge-to-edge.

kkafar commented:
I do not remember from the top of my head the exact interaction there. I'll update the PR when doing testing before the release.

kkafar reviewed (approved):
I've added `react-native-edge-to-edge` to our example apps & added enforced values for `topInsetEnabled` when `EDGE_TO_EDGE===true`.

Did some sanity manual testing & it looks that we're good. I'll merge once the CI passes. Took a while, I know 😅 

Thanks a lot @zoontek.
